datasets/_global/shu_uyghur_companies/crawler.py

```diff
@@ -0,0 +1,156 @@
+            entity.add("notes", row.pop("notes"))
+            entity.add("classification", sheet)
+            entity.add("topics", "export.control")
+            entity.add(
+                "program",
+                "Companies Named in Media and Academic Reports as engaging in Labour Transfers or other XUAR Government Programs",
+            )
+            # entity.add("topics", "export.risk") # not sure if this is needed
+
+            if parent is not None:
+                parent_ent = context.make("Company")
+                parent_ent.id = context.make_id(parent, parent_en)
+                parent_ent.add("name", parent, lang="zhu")
+                parent_ent.add("name", parent_en, lang="eng")
+                parent_ent.add("topics", "export.control")
+                context.emit(parent_ent)
+                entity.add("parent", parent_ent)
+
+            apply_addresses(
+                context,
+                entity,
+                addr=addr_xuar or addr,
+                addr_en=addr_xuar_en or addr_en,
+                city=None,
+                city_en=None,
+            )
+
+            context.emit(entity, target=True)
+
+        context.audit_data(
+            row,
+            ignore=[
+                "add_date",
+                "source_1",
+                "updated",
+                "source",
+            ],
+        )
+
+
+def crawl_operating(context: Context, companies_url):
+    path = context.fetch_resource("companies_registry.xlsx", companies_url[0])
+    workbook: openpyxl.Workbook = openpyxl.load_workbook(path, read_only=True)
+    assert set(workbook.sheetnames) == OPERATING_SHEETS
+    for row in h.parse_xlsx_sheet(
+        context, sheet=workbook["1. Companies Operating in XUAR"], skiprows=1
+    ):
+        name_en = row.pop("company_machine_translated_english")
+        addr = row.pop("address")
+        sector = row.pop("sector")
+
+        entity = context.make("Company")
+        entity.id = context.make_id(name_en, addr, sector)
+        entity.add("name", row.pop("company"), lang="zhu")
+        entity.add("name", name_en, lang="eng")
+        entity.add("sector", sector, lang="zhu")
+        entity.add("sector", row.pop("sector_english"), lang="eng")
+        entity.add("program", "Companies operating in the Uyghur region")
+
+        apply_addresses(
+            context,
+            entity,
+            addr=addr,
+            addr_en=row.pop("address_machine_translated_english"),
+            city=row.pop("city"),
+            city_en=row.pop("city_english"),
+        )
+
+        context.emit(entity)
+
+    context.audit_data(row)
+
+
+def crawl(context: Context):
+    doc = context.fetch_html(context.data_url, cache_days=1)
+    companies_url = doc.xpath(
+        './/a[contains(text(), "Companies Operating in the Uyghur Region")]/@href'
+    )
+    h.assert_url_hash(
+        context, companies_url[0], "e71aab15641be03abeb8b6fce1564aad8714a124"
```

Comment:
What's the rationale for putting a hash on these - are we sure the parser will break?

Reply:
https://github.com/opensanctions/opensanctions/pull/1671#discussion_r1883977497 JD suggested to add it so that we are notified when there's another set of research and update the metadata accordingly